plugins/plugin-machine/che-plugin-machine-ext-client/src/main/java/org/eclipse/che/ide/extension/machine/client/actions/SelectCommandComboBoxReady.java

```diff
@@ -333,17 +315,141 @@ public void onConfigurationsUpdated(CommandConfiguration command) {
     @Override
     public void onWsAgentStarted(WsAgentStateEvent event) {
         loadCommands(null);
+        loadMachines();
     }
 
+    private void loadMachines() {
+        machineServiceClient.getMachines(workspaceId).then(new Operation<List<MachineDto>>() {
+            @Override
+            public void apply(List<MachineDto> machines) throws OperationException {
+                addMachineActions(machines);
+            }
+        });
+    }
+
+
     @Override
     public void onWsAgentStopped(WsAgentStateEvent event) {
+    }
+
+    @Override
+    public void onMachineRunning(MachineStateEvent event) {
+        MachineDto machine = event.getMachine();
 
+        addMachineAction(machine);
     }
 
-    private static class CommandsComparator implements Comparator<CommandConfiguration> {
+    @Override
+    public void onMachineDestroyed(MachineStateEvent event) {
+        MachineDto machine = event.getMachine();
+
+        this.removeMachineAction(machine);
+    }
+
+    private void addMachineActions(List<MachineDto> machines) {
```

Comment:
We agreed not to use DTO in business logic, use `org.eclipse.che.api.core.model.machine.Machine`